### PR TITLE
Componentize the snmp subcommands

### PR DESCRIPTION
**Pull request by FlorianVeaux**
Componentize the "walk" and "scan" SNMP commands. This is to allow registering a new "AgentTaskListener" with remote-config that will later call the RunDeviceScan method of the component

### CODE REVIEW

File: comp/snmpscan/def/component.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package snmpscan ...
```
Comment (GustavoCaso) [inline comp/snmpscan/def/component.go:6]: Could you provide a complete package comment? 😄

File: comp/snmpscan/impl/snmpscan.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+type snmpScannerImpl struct {
+	log   log.Component
+	demux demultiplexer.Component
+}
+
+// func (s snmpScannerImpl) onAgentTaskEvent(taskType rcclienttypes.TaskType, task rcclienttypes.AgentTaskConfig) (bool, error) {
+// 	return false, nil
+// }
```
Comment (GustavoCaso) [inline comp/snmpscan/impl/snmpscan.go:51]: Probably good to remove this until we can add the remote config task event

**Comment by GustavoCaso** [comment]
Left a few suggestions. Also, would be great if we could add test for the new component 😄

File: comp/snmpscan/def/component.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	// fields that aren't part of parse.SNMPConfig
+	SecurityLevel           string
+	UseUnconnectedUDPSocket bool
+}
+
+// Component is the component type.
+type Component interface {
+	// Triggers a device scan
+	RunDeviceScan(snmpConection *gosnmp.GoSNMP, deviceNamespace string) error
+	RunSnmpWalk(snmpConection *gosnmp.GoSNMP, firstOid string) error
```
[the remark this answers is not in the document]
Comment (FlorianVeaux) [reply comp/snmpscan/def/component.go:28]: Yeah I started doing that but it ended up being requiring a much larger refactoring of the code. Passing only SnmpConnectionParams would leave to the code in "options.go" to be used in various places.

That component is very much tied to gosnmp anyway so I thought this was not too much of an issue

File: comp/snmpscan/def/component.go (hunk 1 above)
Comment (FlorianVeaux) [reply comp/snmpscan/def/component.go:6]: Oops yeah missed that with the generated tool!

File: comp/snmpscan/def/component.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply comp/snmpscan/def/component.go:28]: > That component is very much tied to gosnmp anyway so I thought this was not too much of an issue

That is a very good point. 

From my part I'm ok leaving it the way it is currently

File: comp/snmpscan/def/component.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+// team: network-device-monitoring
+
+// Requires defines the dependencies for the snmpscan component
+type Requires struct {
+	fx.In
+	Logger        log.Component
+	Demultiplexer demultiplexer.Component
+}
```
Comment (GustavoCaso) [inline comp/snmpscan/def/component.go:23]: We want to avoid referencing `"go.uber.org/fx"` or internal components requires.

You can refer the documentation for creating components here: https://datadoghq.dev/datadog-agent/components/creating-components/

TL;DR
- Move the `Requires` struct into the `impl` folder.
- replace `fx.In` with `compdef.In` which is a one to one. 

If you want to check how we do it on other components, you can check https://github.com/DataDog/datadog-agent/blob/main/comp/core/healthprobe/impl/healthprobe.go#L27-L28

File: comp/snmpscan/impl/snmpscan.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+// Provides defines the output of the snmpscan component
+type Provides struct {
+	Comp       snmpscan.Component
+	RCListener rcclienttypes.TaskListenerProvider
+}
+
+// NewComponent creates a new snmpscan component
+func NewComponent(reqs snmpscan.Requires) (Provides, error) {
```
Comment (GustavoCaso) [inline comp/snmpscan/impl/snmpscan.go:23]: Once you move the `Requires` struct here, you can reference directly 😄

File: comp/snmpscan/def/component.go (hunk 4 above)
Comment (FlorianVeaux) [reply comp/snmpscan/def/component.go:23]: I've struggled a bit with test, could you maybe help me here?

If I do as you say, my test `TestSnmpScanComp` starts failing, fx is unable to create my component.
It fails with the error `snmpscanimpl.Requires (did you mean to Provide it?)` which is well, what I'm trying to build.

How would you go about fixing that? I've turned a bit in circles here

Comment (FlorianVeaux) [reply comp/snmpscan/def/component.go:23]: Also: I realized there was some component liniting tool which was failing when I tried keeping the struct in `impl` and importing `fx.In`. That's great, it let me now I should import fx in the `impl` folder.

That said, it didn't complain about importing `fx` in the def folder. Is that an oversight with the linter?

File: comp/snmpscan/impl/snmpscan_test.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	"go.uber.org/fx"
+)
+
+func TestSnmpScanComp(t *testing.T) {
+	deps := fxutil.Test[snmpscan.Requires](
+		t,
+		fx.Supply(core.BundleParams{}),
+		demultiplexerimpl.MockModule(),
+		core.MockBundle(),
+	)
```
Comment (GustavoCaso) [inline comp/snmpscan/impl/snmpscan_test.go:23]: Since we are transition to having no fx reference in the `impl` folder we are going through a bit of a migration in which writing test is not straight forward as the ASM would want, sorry about that. 

You could try doing something like: 

```go
import	logmock "github.com/DataDog/datadog-agent/comp/core/log/mock"

type deps struct {
	fx.In
	Demultiplexer demultiplexer.Mock
}


lc := compdef.NewTestLifecycle(t)
logComponent := logmock.New(t)
testDeps := fxutil.Test[deps](t, demultiplexerimpl.MockModule(), defaultforwarder.MockModule(), core.MockBundle())

requires := Requires{
	Logger: logComponent,
	Demultiplexer: testDeps.Demultiplexer
}
```

let me know if you have any doubt 😄

Comment (FlorianVeaux) [reply comp/snmpscan/impl/snmpscan_test.go]: Oh indeed this works, I though importing fx.In in the test would also fail but looks like it's fine because it's a test?
Seems to be working now, thanks!

**Review by TCheruy — APPROVED** [review]

**Review by GustavoCaso — COMMENTED** [review]
Great work on the component side 🎉 

Thanks for taking into account all the suggested feedback. I left one last question regarding testing.

File: cmd/agent/subcommands/snmp/command.go [hunk 7]
```diff
@@ -413,61 +420,18 @@ func scanDevice(connParams *connectionParams, args argsType, conf config.Compone
...
-func gatherPDUs(snmp *gosnmp.GoSNMP) ([]*gosnmp.SnmpPDU, error) {
-	var pdus []*gosnmp.SnmpPDU
-	err := gosnmplib.ConditionalWalk(snmp, "", false, func(dataUnit gosnmp.SnmpPDU) (string, error) {
-		pdus = append(pdus, &dataUnit)
-		return gosnmplib.SkipOIDRowsNaive(dataUnit.Name), nil
-	})
+	err = snmpScanner.RunDeviceScan(snmp, namespace)
 	if err != nil {
-		return nil, err
+		fmt.Fprintf(os.Stderr, "Unable to perform device scan: %v\n", err)
```
Comment (dplepage-dd) [inline cmd/agent/subcommands/snmp/command.go:428]: How are you deciding which errors should be returned and which should just be printed? It feels like since scanDevice can return an error it should do so instead of printing to stderr and returning nil.

File: comp/snmpscan/impl/snmpscan_test.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+	snmpConnection := gosnmp.Default
+	snmpConnection.LocalAddr = "127.0.0.1"
+	snmpConnection.Port = 0
+
+	err = snmpScanner.Comp.RunDeviceScan(snmpConnection, "default")
+	assert.ErrorContains(t, err, "&GoSNMP.Conn is missing. Provide a connection or use Connect()")
+
+	err = snmpScanner.Comp.RunDeviceScan(snmpConnection, "default")
+	assert.ErrorContains(t, err, "&GoSNMP.Conn is missing. Provide a connection or use Connect()")
```
[the remark this answers is not in the document]
Comment (dplepage-dd) [reply comp/snmpscan/impl/snmpscan_test.go:47]: This test will also fail if gosnmp ever changes their error text, which isn't great, but I guess if GoSNMP isn't returning typed errors there's not a lot we can do about that.

File: comp/snmpscan/impl/walkdevice.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+			fmt.Print("Hex-STRING: " + strings.Join(strBytes, " ") + "\n")
+		} else {
+			fmt.Printf("STRING: %s\n", string(b))
+		}
+	case gosnmp.ObjectIdentifier:
+		fmt.Printf("OID: %s\n", pdu.Value)
+	case gosnmp.TimeTicks:
+		fmt.Print(pdu.Value, "\n")
+	case gosnmp.Counter32:
+		fmt.Printf("Counter32: %d\n", pdu.Value.(uint))
```
Comment (ogaca-dd) [inline comp/snmpscan/impl/walkdevice.go:48]: Did you consider using `%v` to avoid casting for each type?

Comment (FlorianVeaux) [reply comp/snmpscan/impl/walkdevice.go:48]: Not really, this PR is moving code around so I'd like to minimize changes to the existing code to keep the PR small although there could be a few improvements to make indeed

File: cmd/agent/subcommands/snmp/command.go (hunk 7 above)
Comment (FlorianVeaux) [reply cmd/agent/subcommands/snmp/command.go:428]: This one is a mistake, and it should return an error indeed

Comment (FlorianVeaux) [reply cmd/agent/subcommands/snmp/command.go:428]: I've kept the logic similar for the rest, the only place where we log the error is when we cannot contact the Agent because this is recoverable error (if there are enough arguments)

File: comp/snmpscan/impl/devicescan.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/DataDog/datadog-agent/pkg/networkdevice/metadata"
+	"github.com/DataDog/datadog-agent/pkg/snmp/gosnmplib"
+	"github.com/gosnmp/gosnmp"
+)
+
+func (s snmpScannerImpl) RunDeviceScan(snmpConnection *gosnmp.GoSNMP, deviceNamespace string) error {
+	forwarder, err := s.demux.GetEventPlatformForwarder()
+	if err != nil {
+		return fmt.Errorf("unable to get sender: %w", err)
+	}
```
Comment (GustavoCaso) [inline comp/snmpscan/impl/devicescan.go:24]: We could move this code to the component initialize function and return an error the event platform is not properly initialized. That would catch issue earlier, and it would avoid having to call this function everytime we execute the device scan command.

```go
func NewComponent(reqs Requires) (Provides, error) {
	forwarder, err := s.demux.GetEventPlatformForwarder()
	if err != nil {
		return Provides{}, err
	}
	scanner := snmpScannerImpl{
		log:   reqs.Logger,
		forwarder: forwarder,
	}
	provides := Provides{
		Comp: scanner,
	}
	return provides, nil
}
```

Comment (FlorianVeaux) [reply comp/snmpscan/impl/devicescan.go:24]: Great idea!

**Review by GustavoCaso — APPROVED** [review]
Amazing work 🎉
